File: backend-python/backend-saude/app/controller/pontoController.py
# Importações de bibliotecas necessárias
from datetime import datetime, timedelta  # Manipulação de datas e horários
from flask import jsonify, request        # Utilidades Flask para requisição e resposta
from app.db.database import get_db_connection  # Função de conexão com o banco de dados
from app.services.biometric import IndexSearch, identify_user  # Lógica biométrica
import requests  # Para fazer chamadas HTTP ao backend em Node.js


# http://localhost:3001

# http://biometrico.itaguai.rj.gov.br:3001

# ===========================
# Função auxiliar para envio de e-mail
# ===========================
def send_email(subject, recipient, body):
    try:
        # Envia um POST com os dados do e-mail para o backend Node.js
        response = requests.post("http://biometrico.itaguai.rj.gov.br:3001/api/enviar-email", json={
            "subject": subject,
            "recipient": recipient,
            "body": body
        })
        response.raise_for_status()  # Lança exceção se o status não for 2xx
        print(f"E-mail enviado para {recipient}")
    except requests.RequestException as e:
        print(f"Erro ao enviar e-mail: {e}")

# ===========================
# Função principal: Registrar ponto via biometria
# ===========================
def register_ponto():
    data = request.json or {}  # Lê os dados enviados no corpo da requisição

    # Captura os parâmetros principais enviados pelo terminal
    unidade_id_terminal = data.get('unidade_id')
    data_registro = data.get('data') or datetime.now().date().strftime("%Y-%m-%d")
    hora_entrada = data.get('hora_entrada') or datetime.now().strftime("%H:%M:%S")

    # ===========================
    # 1. Preparação para identificação biométrica
    # ===========================
    IndexSearch.ClearDB()  # Limpa base de digitais temporária
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT id_biometrico, id FROM funcionarios")
    for row in cursor.fetchall():
        IndexSearch.AddFIR(row[0], int(row[1]))  # Adiciona cada funcionário à base de identificação
    conn.close()

    # ===========================
    # 2. Captura da digital no leitor e identificação
    # ===========================
    fir_data = identify_user()
    if not fir_data:
        return jsonify({"message": "Nenhuma impressão digital capturada. Por favor, tente novamente."}), 400

    IndexSearch.IdentifyUser(fir_data, 5)  # Tenta identificar com tolerância 5
    if IndexSearch.UserID == 0:
        return jsonify({"message": "Usuário não identificado. Digital não cadastrada no sistema."}), 401

    funcionario_id = IndexSearch.UserID  # ID do funcionário identificado

    # ===========================
    # 3. Buscar dados do funcionário no banco
    # ===========================
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("""
        SELECT id, nome, cpf, unidade_id, matricula, cargo, id_biometrico, email
        FROM funcionarios WHERE id = %s
    """, (funcionario_id,))
    user_data = cursor.fetchone()
    if not user_data:
        return jsonify({"message": "Funcionário não encontrado no banco de dados."}), 404

    # Atribui as informações do funcionário a variáveis
    funcionario_id, user_name, cpf, unidade_id_funcionario, matricula, cargo, id_biometrico, email = user_data

    # ===========================
    # 4. Validação de unidade (terminal vs funcionário)
    # ===========================
    if not unidade_id_terminal:
        return jsonify({"message": "unidade_id é obrigatório"}), 400

    if unidade_id_funcionario != unidade_id_terminal:
        # Busca nomes das unidades para detalhar o erro
        cursor.execute("SELECT nome FROM unidades WHERE id = %s", (unidade_id_funcionario,))
        unidade_funcionario_nome = cursor.fetchone()
        cursor.execute("SELECT nome FROM unidades WHERE id = %s", (unidade_id_terminal,))
        unidade_terminal_nome = cursor.fetchone()

        print(f"[ACESSO NEGADO] Funcionário: {user_name} (ID: {funcionario_id}) | Unidade do funcionário: {unidade_funcionario_nome[0] if unidade_funcionario_nome else unidade_id_funcionario} | Unidade do terminal: {unidade_terminal_nome[0] if unidade_terminal_nome else unidade_id_terminal} | Data/Hora: {datetime.now().strftime('%d/%m/%Y %H:%M:%S')}")

        return jsonify({
            "message": "Funcionário não pertence a esta unidade.",
            "funcionario": user_name,
            "unidade_funcionario": unidade_funcionario_nome[0] if unidade_funcionario_nome else unidade_id_funcionario,
            "unidade_terminal": unidade_terminal_nome[0] if unidade_terminal_nome else unidade_id_terminal
        }), 403

    # ===========================
    # 5. Verifica se o funcionário está de férias
    # ===========================
    data_atual = datetime.strptime(data_registro, "%Y-%m-%d").date()
    cursor.execute("""
        SELECT data_inicio, data_fim FROM ferias 
        WHERE funcionario_id = %s AND data_inicio <= %s AND data_fim >= %s
    """, (funcionario_id, data_atual, data_atual))
    ferias_data = cursor.fetchone()
    if ferias_data:
        return jsonify({"message": "Funcionário de férias, você não pode registrar o ponto!"}), 400

    # ===========================
    # 6. Busca escala do funcionário e verifica registros considerando a escala
    # ===========================
    cursor.execute("SELECT tipo_escala FROM funcionarios WHERE id = %s", (funcionario_id,))
    escala_result = cursor.fetchone()
    escala = escala_result[0] if escala_result else '8h'
    
    # Para escalas de 24h, verifica registros dos últimos 2 dias
    # Para outras escalas, verifica apenas o dia atual
    if escala in ['24h', '24x72']:
        cursor.execute("""
            SELECT id, hora_entrada, hora_saida, data_hora FROM registros_ponto 
            WHERE funcionario_id = %s 
            AND DATE(data_hora) >= %s - INTERVAL '1 day'
            AND DATE(data_hora) <= %s
            AND hora_saida IS NULL
            ORDER BY data_hora DESC LIMIT 1
        """, (funcionario_id, data_atual, data_atual))
        ultimo_ponto_pendente = cursor.fetchone()
        
        # Também verifica se já tem registro completo no dia atual
        cursor.execute("""
            SELECT id, hora_entrada, hora_saida, data_hora FROM registros_ponto 
            WHERE funcionario_id = %s 
            AND DATE(data_hora) = %s 
            ORDER BY data_hora DESC LIMIT 1
        """, (funcionario_id, data_atual))
        ultimo_ponto_hoje = cursor.fetchone()
    else:
        # Para escalas normais, mantém a lógica atual
        cursor.execute("""
            SELECT id, hora_entrada, hora_saida, data_hora FROM registros_ponto 
            WHERE funcionario_id = %s 
            AND DATE(data_hora) = %s 
            ORDER BY data_hora DESC LIMIT 1
        """, (funcionario_id, data_atual))
        ultimo_ponto_hoje = cursor.fetchone()
        ultimo_ponto_pendente = ultimo_ponto_hoje if ultimo_ponto_hoje and ultimo_ponto_hoje[2] is None else None

    mensagem = ""

    # ===========================
    # 7. REGISTRO DE ENTRADA
    # ===========================
    # Se não há ponto pendente E não há registro completo hoje
    if not ultimo_ponto_pendente and (not ultimo_ponto_hoje or ultimo_ponto_hoje[2] is None):
        # Não há registro pendente, registra entrada
        payload = {
            "funcionario_id": funcionario_id,
            "unidade_id": unidade_id_terminal,
            "data": data_registro,
            "hora_entrada": hora_entrada,
            "hora_saida": None,
            "id_biometrico": id_biometrico
        }
        try:
            response = requests.post(
                "http://biometrico.itaguai.rj.gov.br:3001/reg/calcular-registro-ponto",
                json=payload,
                timeout=10
            )
            response.raise_for_status()
        except requests.RequestException as e:
            # Se o Node.js retornou erro, tente pegar a mensagem do corpo da resposta
            if e.response is not None and e.response.content:
                try:
                    error_json = e.response.json()
                    return jsonify(error_json), e.response.status_code
                except Exception:
                    # Se não for JSON, retorna texto puro
                    return jsonify({"message": e.response.text}), e.response.status_code
            # Se não tem resposta do Node, retorna erro genérico
            return jsonify({"message": "Erro ao registrar ponto no sistema"}), 500

        # Envia e-mail de comprovante de entrada
        data_hora = datetime.now()
        mensagem = (
            f"Registro de entrada realizado com sucesso para funcionario: {user_name}\n"
            f"Comprovante enviado para o e-mail {email}"
        )
        send_email(
            subject="Registro de Entrada - Ponto Registrado",
            recipient=email,
            body=f"""
                Prezado(a) {user_name},

                Este e-mail confirma o registro de seu ponto conforme as informações abaixo:

                Entrada registrada com sucesso.

                Profissional: {user_name}
                Data/Hora: {data_hora.strftime('%d/%m/%Y %H:%M:%S')}

                Se precisar de suporte ou tiver dúvidas, entre em contato com a Prefeitura de Itaguaí.

                Atenciosamente,
                Prefeitura de Itaguaí
            """
        )
        print(f"[ENTRADA REGISTRADA] Funcionário: {user_name} (ID: {funcionario_id}) | Unidade: {unidade_id_terminal} | Data/Hora: {data_registro} {hora_entrada}")

    # ===========================
    # 8. REGISTRO DE SAÍDA (com bloqueio de 5 minutos)
    # ===========================
    elif ultimo_ponto_pendente and ultimo_ponto_pendente[2] is None:
        # Há entrada sem saída (pode ser de ontem para escalas 24h), registra saída
        hora_entrada_time = ultimo_ponto_pendente[1]
        data_entrada = datetime.combine(ultimo_ponto_pendente[3].date(), hora_entrada_time)
        tempo_decorrido = datetime.now() - data_entrada
        tempo_decorrido_minutos = tempo_decorrido.total_seconds() / 60

        if tempo_decorrido_minutos < 1:
            tempo_restante = int(1 - tempo_decorrido_minutos) + 1
            print(f"[TENTATIVA BLOQUEADA] Funcionário: {user_name} (ID: {funcionario_id}) | Tempo decorrido: {tempo_decorrido_minutos:.2f} minutos | Tempo restante: {tempo_restante} minuto(s)")
            return jsonify({
                "message": f"Você deve aguardar pelo menos 5 minutos após a entrada para registrar a saída. Tempo restante: {tempo_restante} minuto(s)."
            }), 400

        # Se passou mais de 5 minutos, registra a saída
        hora_saida = datetime.now().strftime("%H:%M:%S")
        payload = {
            "funcionario_id": funcionario_id,
            "unidade_id": unidade_id_terminal,
            "data": ultimo_ponto_pendente[3].date().strftime("%Y-%m-%d"),  # Usa a data da entrada
            "hora_entrada": hora_entrada_time.strftime("%H:%M:%S"),
            "hora_saida": hora_saida,
            "id_biometrico": id_biometrico
        }
        try:
            response = requests.post(
                "http://biometrico.itaguai.rj.gov.br:3001/reg/calcular-registro-ponto",
                json=payload,
                timeout=10
            )
            response.raise_for_status()
        except requests.RequestException as e:
            # Se o Node.js retornou erro, tente pegar a mensagem do corpo da resposta
            if e.response is not None and e.response.content:
                try:
                    error_json = e.response.json()
                    return jsonify(error_json), e.response.status_code
                except Exception:
                    # Se não for JSON, retorna texto puro
                    return jsonify({"message": e.response.text}), e.response.status_code
            # Se não tem resposta do Node, retorna erro genérico
            return jsonify({"message": "Erro ao registrar ponto no sistema"}), 500

        # Envia e-mail de comprovante de saída
        mensagem = (
            f"Registro de saida realizado com sucesso para funcionario: {user_name}\n"
            f"Comprovante enviado para o e-mail {email}"
        )
        send_email(
            subject="Registro de Saída - Ponto Registrado",
            recipient=email,
            body=f"""
                Prezado(a) {user_name},

                Este e-mail confirma o registro de sua saída conforme as informações abaixo:

                Saída registrada com sucesso.

                Profissional: {user_name}
                Data/Hora: {datetime.now().strftime('%d/%m/%Y %H:%M:%S')}

                Se precisar de suporte ou tiver dúvidas, entre em contato com a Prefeitura de Itaguaí.

                Atenciosamente,
                Prefeitura de Itaguaí
            """
        )
        print(f"[SAÍDA REGISTRADA] Funcionário: {user_name} (ID: {funcionario_id}) | Unidade: {unidade_id_terminal} | Tempo trabalhado: {tempo_decorrido_minutos:.2f} minutos")

    # ===========================
    # 9. Caso já tenha registro completo hoje
    # ===========================
    elif ultimo_ponto_hoje and ultimo_ponto_hoje[1] is not None and ultimo_ponto_hoje[2] is not None:
        # Já tem entrada e saída hoje
        return jsonify({
            "message": f"Você já bateu sua saída hoje ({data_atual.strftime('%d/%m/%Y')})."
        }), 400

    # ===========================
    # 10. Caso inesperado
    # ===========================
    else:
        return jsonify({
            "message": "Estado do registro de ponto não identificado. Contate o suporte."
        }), 500

    # Finaliza conexões com banco
    cursor.close()
    conn.close()

    # Determina o tipo baseado na condição que foi executada
    tipo_registro = "entrada" if not ultimo_ponto_pendente and (not ultimo_ponto_hoje or ultimo_ponto_hoje[2] is None) else "saida"

    # Resposta de sucesso com detalhes do registro
    return jsonify({
        "message": mensagem,
        "funcionario": user_name,
        "matricula": matricula,
        "data_hora": f"{data_registro} {hora_entrada}",
        "tipo": tipo_registro
    }), 200

File: backend-java/src/main/java/br/com/bio/biomtria/controller/pontoController.java
package br.com.bio.biomtria.controller;

import br.com.bio.biomtria.service.BiometricService;
// import br.com.bio.biomtria.jdbcTemplate;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.SimpleDateFormat;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;
import java.util.Map;

/**
 * Controlador para as operações de ponto eletrônico
 * A rota não é definida aqui, mas sim em pontoRoutes.java
 */
@Component
public class pontoController {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Autowired
    private BiometricService biometricService;

    @Autowired
    private RestTemplate restTemplate;

    private static final String EMAIL_API_URL = "http://biometrico.itaguai.rj.gov.br:3001/api/enviar-email";
    private static final String PONTO_API_URL = "http://biometrico.itaguai.rj.gov.br:3001/reg/calcular-registro-ponto";

    /**
     * Método para cadastrar (enroll) uma impressão digital
     * 
     * @param idBiometrico Identificador biométrico a ser associado à digital
     * @return String contendo os dados da impressão digital no formato
     *         TextEncodeFIR
     */
    public ResponseEntity<?> enrollUser(String idBiometrico) {
        try {
            System.out.println("Iniciando cadastro de digital para ID: " + idBiometrico);

            // Captura e exporta o template biométrico real
            byte[] template = biometricService.enrollUserTemplate();

            if (template == null || template.length == 0) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(Map.of("message", "Falha ao cadastrar impressão digital. Tente novamente."));
            }

            // Salva no banco de dados (campo binário)
            try {
                jdbcTemplate.update(
                        "UPDATE funcionarios SET id_biometrico = ? WHERE id = ?",
                        template, idBiometrico);
            } catch (Exception e) {
                System.out.println("Aviso: Não foi possível salvar a digital no banco: " + e.getMessage());
                // Continua mesmo que não consiga salvar no banco
            }

            Map<String, Object> response = new HashMap<>();
            response.put("message", "Impressão digital cadastrada com sucesso!");
            response.put("id_biometrico", idBiometrico);
            response.put("template_length", template.length);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Erro ao cadastrar impressão digital: " + e.getMessage()));
        }
    }

    /**
     * Método auxiliar para envio de e-mail
     */
    private void sendEmail(String subject, String recipient, String body) {
        try {
            Map<String, String> emailRequest = new HashMap<>();
            emailRequest.put("subject", subject);
            emailRequest.put("recipient", recipient);
            emailRequest.put("body", body);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            HttpEntity<Map<String, String>> entity = new HttpEntity<>(emailRequest, headers);

            restTemplate.postForEntity(EMAIL_API_URL, entity, String.class);
            System.out.println("E-mail enviado para " + recipient);
        } catch (Exception e) {
            System.out.println("Erro ao enviar e-mail: " + e.getMessage());
        }
    }

    /**
     * Método principal para registro de ponto via biometria
     * A rota é definida na classe pontoRoutes
     */
    public ResponseEntity<?> registerPonto(Map<String, Object> requestData) {
        // ===========================
        // 1. Preparação e validação dos parâmetros
        // ===========================

        // Captura os parâmetros principais enviados pelo terminal
        Object unidadeIdObj = requestData.get("unidade_id");
        Integer unidadeIdTerminal = null;
        if (unidadeIdObj != null) {
            if (unidadeIdObj instanceof Integer) {
                unidadeIdTerminal = (Integer) unidadeIdObj;
            } else {
                try {
                    unidadeIdTerminal = Integer.parseInt(unidadeIdObj.toString());
                } catch (NumberFormatException e) {
                    return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                            .body(Map.of("message", "unidade_id inválido: " + unidadeIdObj));
                }
            }
        }

        // Formata a data atual se não for fornecida
        String dataRegistro;
        if (requestData.get("data") != null) {
            dataRegistro = (String) requestData.get("data");
        } else {
            dataRegistro = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
        }

        // Formata a hora atual se não for fornecida
        String horaEntrada;
        if (requestData.get("hora_entrada") != null) {
            horaEntrada = (String) requestData.get("hora_entrada");
        } else {
            horaEntrada = LocalTime.now().format(DateTimeFormatter.ofPattern("HH:mm:ss"));
        }

        // ===========================
        // 2. Prepara base biométrica temporária
        // ===========================
        try {
            // Limpa base de digitais temporária
            biometricService.clearIndexSearch();

            // Carrega todas as digitais do banco para identificação
            jdbcTemplate.query(
                    "SELECT id_biometrico, id FROM funcionarios",
                    (rs, rowNum) -> {
                        biometricService.addFIRToIndex(rs.getString("id_biometrico"), rs.getInt("id"));
                        return null;
                    });
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Erro ao preparar base biométrica: " + e.getMessage()));
        }
        // ===========================
        // 3. Captura e identificação biométrica
        // ===========================
        System.out.println("[DEBUG] Iniciando identificação biométrica...");
        try {
            int funcionarioId = biometricService.identifyUserByTemplate(jdbcTemplate);
            if (funcionarioId == 0) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                        .body(Map.of("message", "Digital não reconhecida. Por favor, tente novamente."));
            }

            // ===========================
            // 3. Buscar dados do funcionário no banco
            // ===========================
            Map<String, Object> userData = jdbcTemplate.queryForMap(
                    "SELECT id, nome, cpf, unidade_id, matricula, cargo, id_biometrico, email " +
                            "FROM funcionarios WHERE id = ?",
                    funcionarioId);

            if (userData == null || userData.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("message", "Funcionário não encontrado no banco de dados."));
            }

            // Atribui as informações do funcionário a variáveis
            String userName = userData.get("nome") != null ? userData.get("nome").toString() : null;
            Integer unidadeIdFuncionario = userData.get("unidade_id") instanceof Integer
                    ? (Integer) userData.get("unidade_id")
                    : (userData.get("unidade_id") != null ? Integer.parseInt(userData.get("unidade_id").toString())
                            : null);
            String matricula = userData.get("matricula") != null ? userData.get("matricula").toString() : null;
            String idBiometrico = userData.get("id_biometrico") != null ? userData.get("id_biometrico").toString()
                    : null;
            String email = userData.get("email") != null ? userData.get("email").toString() : null;

            // ===========================
            // 4. Validação de unidade (terminal vs funcionário)
            // ===========================
            if (unidadeIdTerminal == null) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(Map.of("message", "unidade_id é obrigatório"));
            }

            if (!unidadeIdFuncionario.equals(unidadeIdTerminal)) {
                // Busca nomes das unidades para detalhar o erro
                String unidadeFuncionarioNome = jdbcTemplate.queryForObject(
                        "SELECT nome FROM unidades WHERE id = ?",
                        String.class,
                        unidadeIdFuncionario);

                String unidadeTerminalNome = jdbcTemplate.queryForObject(
                        "SELECT nome FROM unidades WHERE id = ?",
                        String.class,
                        unidadeIdTerminal);

                System.out.printf("[ACESSO NEGADO] Funcionário: %s (ID: %d) | Unidade do funcionário: %s | " +
                        "Unidade do terminal: %s | Data/Hora: %s%n",
                        userName, funcionarioId,
                        unidadeFuncionarioNome != null ? unidadeFuncionarioNome : unidadeIdFuncionario,
                        unidadeTerminalNome != null ? unidadeTerminalNome : unidadeIdTerminal,
                        LocalDateTime.now().format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss")));

                Map<String, Object> response = new HashMap<>();
                response.put("message", "Funcionário não pertence a esta unidade.");
                response.put("funcionario", userName);
                response.put("unidade_funcionario",
                        unidadeFuncionarioNome != null ? unidadeFuncionarioNome : unidadeIdFuncionario);
                response.put("unidade_terminal", unidadeTerminalNome != null ? unidadeTerminalNome : unidadeIdTerminal);

                return ResponseEntity.status(HttpStatus.FORBIDDEN).body(response);
            }

            // ===========================
            // 5. Verifica se o funcionário está de férias
            // ===========================
            LocalDate dataAtual = LocalDate.parse(dataRegistro);
            Integer feriasCount = jdbcTemplate.queryForObject(
                    "SELECT COUNT(*) FROM ferias WHERE funcionario_id = ? AND data_inicio <= ? AND data_fim >= ?",
                    Integer.class,
                    funcionarioId, dataAtual, dataAtual);

            if (feriasCount != null && feriasCount > 0) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(Map.of("message", "Funcionário de férias, você não pode registrar o ponto!"));
            }

            // ===========================
            // 6. Busca escala do funcionário e verifica registros considerando a escala
            // ===========================
            String escala = jdbcTemplate.queryForObject(
                    "SELECT tipo_escala FROM funcionarios WHERE id = ?",
                    String.class,
                    funcionarioId);

            if (escala == null) {
                escala = "8h"; // Escala padrão se não estiver definida
            }

            Map<String, Object> ultimoPontoPendente = null;
            Map<String, Object> ultimoPontoHoje = null;

            // Para escalas de 24h, verifica registros dos últimos 2 dias
            if ("24h".equals(escala) || "24x72".equals(escala)) {
                // Busca último ponto pendente (sem hora de saída) dos últimos 2 dias
                try {
                    ultimoPontoPendente = jdbcTemplate.queryForMap(
                            "SELECT id, hora_entrada, hora_saida, data_hora FROM registros_ponto " +
                                    "WHERE funcionario_id = ? " +
                                    "AND DATE(data_hora) >= ? - INTERVAL '1 day' " +
                                    "AND DATE(data_hora) <= ? " +
                                    "AND hora_saida IS NULL " +
                                    "ORDER BY data_hora DESC LIMIT 1",
                            funcionarioId, dataAtual, dataAtual);
                } catch (Exception e) {
                    // Nenhum registro encontrado
                }

                // Também verifica se já tem registro completo no dia atual
                try {
                    ultimoPontoHoje = jdbcTemplate.queryForMap(
                            "SELECT id, hora_entrada, hora_saida, data_hora FROM registros_ponto " +
                                    "WHERE funcionario_id = ? " +
                                    "AND DATE(data_hora) = ? " +
                                    "ORDER BY data_hora DESC LIMIT 1",
                            funcionarioId, dataAtual);
                } catch (Exception e) {
                    // Nenhum registro encontrado
                }
            } else {
                // Para escalas normais, mantém a lógica atual
                try {
                    ultimoPontoHoje = jdbcTemplate.queryForMap(
                            "SELECT id, hora_entrada, hora_saida, data_hora FROM registros_ponto " +
                                    "WHERE funcionario_id = ? " +
                                    "AND DATE(data_hora) = ? " +
                                    "ORDER BY data_hora DESC LIMIT 1",
                            funcionarioId, dataAtual);

                    // Se tem ponto hoje e hora_saida é nula, é um ponto pendente
                    if (ultimoPontoHoje != null && ultimoPontoHoje.get("hora_saida") == null) {
                        ultimoPontoPendente = ultimoPontoHoje;
                    }
                } catch (Exception e) {
                    // Nenhum registro encontrado
                }
            }

            String mensagem = "";

            // ===========================
            // 7. REGISTRO DE ENTRADA
            // ===========================
            // Se não há ponto pendente E não há registro completo hoje
            if (ultimoPontoPendente == null &&
                    (ultimoPontoHoje == null || ultimoPontoHoje.get("hora_saida") == null)) {

                // Não há registro pendente, registra entrada
                Map<String, Object> payload = new HashMap<>();
                payload.put("funcionario_id", funcionarioId);
                payload.put("unidade_id", unidadeIdTerminal);
                payload.put("data", dataRegistro);
                payload.put("hora_entrada", horaEntrada);
                payload.put("hora_saida", null);
                payload.put("id_biometrico", idBiometrico);

                try {
                    HttpHeaders headers = new HttpHeaders();
                    headers.setContentType(MediaType.APPLICATION_JSON);
                    HttpEntity<Map<String, Object>> entity = new HttpEntity<>(payload, headers);

                    // Envia para o sistema de ponto
                    restTemplate.postForEntity(
                            PONTO_API_URL,
                            entity,
                            String.class);
                } catch (Exception e) {
                    return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                            .body(Map.of("message", "Erro ao registrar ponto no sistema: " + e.getMessage()));
                }

                // Envia e-mail de comprovante de entrada
                LocalDateTime dataHora = LocalDateTime.now();
                mensagem = String.format(
                        "Registro de entrada realizado com sucesso para funcionario: %s\n" +
                                "Comprovante enviado para o e-mail %s",
                        userName, email);

                sendEmail(
                        "Registro de Entrada - Ponto Registrado",
                        email,
                        String.format(
                                "Prezado(a) %s,\n\n" +
                                        "Este e-mail confirma o registro de seu ponto conforme as informações abaixo:\n\n"
                                        +
                                        "Entrada registrada com sucesso.\n\n" +
                                        "Profissional: %s\n" +
                                        "Data/Hora: %s\n\n" +
                                        "Se precisar de suporte ou tiver dúvidas, entre em contato com a Prefeitura de Itaguaí.\n\n"
                                        +
                                        "Atenciosamente,\n" +
                                        "Prefeitura de Itaguaí",
                                userName, userName,
                                dataHora.format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss"))));

                System.out.printf("[ENTRADA REGISTRADA] Funcionário: %s (ID: %d) | Unidade: %d | Data/Hora: %s %s%n",
                        userName, funcionarioId, unidadeIdTerminal, dataRegistro, horaEntrada);

                // ===========================
                // 8. REGISTRO DE SAÍDA (com bloqueio de 5 minutos)
                // ===========================
            } else if (ultimoPontoPendente != null && ultimoPontoPendente.get("hora_saida") == null) {
                // Há entrada sem saída (pode ser de ontem para escalas 24h), registra saída
                LocalTime horaEntradaTime = LocalTime.parse(ultimoPontoPendente.get("hora_entrada").toString());
                LocalDateTime dataEntrada = LocalDateTime.of(
                        ((java.sql.Timestamp) ultimoPontoPendente.get("data_hora")).toLocalDateTime().toLocalDate(),
                        horaEntradaTime);

                LocalDateTime agora = LocalDateTime.now();
                long tempoDecorridoMinutos = ChronoUnit.MINUTES.between(dataEntrada, agora);

                if (tempoDecorridoMinutos < 5) {
                    int tempoRestante = (int) (5 - tempoDecorridoMinutos);
                    System.out.printf(
                            "[TENTATIVA BLOQUEADA] Funcionário: %s (ID: %d) | Tempo decorrido: %.2f minutos | Tempo restante: %d minuto(s)%n",
                            userName, funcionarioId, (float) tempoDecorridoMinutos, tempoRestante);

                    return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                            .body(Map.of("message", String.format(
                                    "Você deve aguardar pelo menos 5 minutos após a entrada para registrar a saída. Tempo restante: %d minuto(s).",
                                    tempoRestante)));
                }

                // Se passou mais de 5 minutos, registra a saída
                String horaSaida = LocalTime.now().format(DateTimeFormatter.ofPattern("HH:mm:ss"));
                LocalDate dataEntradaDate = ((java.sql.Timestamp) ultimoPontoPendente.get("data_hora"))
                        .toLocalDateTime().toLocalDate();

                Map<String, Object> payload = new HashMap<>();
                payload.put("funcionario_id", funcionarioId);
                payload.put("unidade_id", unidadeIdTerminal);
                payload.put("data", dataEntradaDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))); // Usa a data da
                                                                                                        // entrada
                payload.put("hora_entrada", horaEntradaTime.format(DateTimeFormatter.ofPattern("HH:mm:ss")));
                payload.put("hora_saida", horaSaida);
                payload.put("id_biometrico", idBiometrico);

                try {
                    HttpHeaders headers = new HttpHeaders();
                    headers.setContentType(MediaType.APPLICATION_JSON);
                    HttpEntity<Map<String, Object>> entity = new HttpEntity<>(payload, headers);

                    // Envia para o sistema de ponto
                    restTemplate.postForEntity(
                            PONTO_API_URL,
                            entity,
                            String.class);
                } catch (Exception e) {
                    return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                            .body(Map.of("message", "Erro ao registrar ponto no sistema: " + e.getMessage()));
                }

                // Envia e-mail de comprovante de saída
                mensagem = String.format(
                        "Registro de saída realizado com sucesso para funcionario: %s\n" +
                                "Comprovante enviado para o e-mail %s",
                        userName, email);

                sendEmail(
                        "Registro de Saída - Ponto Registrado",
                        email,
                        String.format(
                                "Prezado(a) %s,\n\n" +
                                        "Este e-mail confirma o registro de sua saída conforme as informações abaixo:\n\n"
                                        +
                                        "Saída registrada com sucesso.\n\n" +
                                        "Profissional: %s\n" +
                                        "Data/Hora: %s\n\n" +
                                        "Se precisar de suporte ou tiver dúvidas, entre em contato com a Prefeitura de Itaguaí.\n\n"
                                        +
                                        "Atenciosamente,\n" +
                                        "Prefeitura de Itaguaí",
                                userName, userName,
                                LocalDateTime.now().format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss"))));

                System.out.printf(
                        "[SAÍDA REGISTRADA] Funcionário: %s (ID: %d) | Unidade: %d | Tempo trabalhado: %.2f minutos%n",
                        userName, funcionarioId, unidadeIdTerminal, (float) tempoDecorridoMinutos);

                // ===========================
                // 9. Caso já tenha registro completo hoje
                // ===========================
            } else if (ultimoPontoHoje != null &&
                    ultimoPontoHoje.get("hora_entrada") != null &&
                    ultimoPontoHoje.get("hora_saida") != null) {

                // Já tem entrada e saída hoje
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(Map.of("message", String.format(
                                "Você já bateu sua saída hoje (%s).",
                                dataAtual.format(DateTimeFormatter.ofPattern("dd/MM/yyyy")))));

                // ===========================
                // 10. Caso inesperado
                // ===========================
            } else {
                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                        .body(Map.of("message", "Estado do registro de ponto não identificado. Contate o suporte."));
            }

            // Determina o tipo baseado na condição que foi executada
            String tipoRegistro = ultimoPontoPendente == null &&
                    (ultimoPontoHoje == null || ultimoPontoHoje.get("hora_saida") == null) ? "entrada" : "saida";

            // Resposta de sucesso com detalhes do registro
            Map<String, Object> responseData = new HashMap<>();
            responseData.put("message", mensagem);
            responseData.put("funcionario", userName);
            responseData.put("matricula", matricula);
            responseData.put("data_hora", dataRegistro + " " + horaEntrada);
            responseData.put("tipo", tipoRegistro);

            return ResponseEntity.ok(responseData);

        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Erro interno: " + e.getMessage()));
        }
    }
}
